// contexts/currentUser.js
import React ,{createContext, useReducer} from 'react'

const initialState = {
    isLoading: false,
    isLogged: null,
    currentUser: null,
    currentUserName: null
}

const reducer = (state, action) => {
    switch (action.type) {
        case 'LOADING':
            return {...state, isLoading: true}
        case 'SET_AUTHORIZED':
            return {
                ...state,
                isLoading:false,
                isLogged:true,
                currentUser: action.payload,
                currentUserName: action.payload.username
            }

        case 'SET_UNAUTHORIZED':
            return {
                ...state,               
                isLogged:false               

            }
            default:
                return state
    }
}

export const CurrentUserContext = createContext()

export const CurrentUserProvider = ({children}) => {
    const value = useReducer(reducer,initialState)
   
    return (
        <CurrentUserContext.Provider value={value}>
        {children}
        </CurrentUserContext.Provider>
    )
        
}
